### PR TITLE
[3.3.5] Core/Entities: Template Optimizer for Locales

### PR DESCRIPTION
**Pull request by Langerz82**
**Changes proposed:**
What if we created the users default locale QueryData, and created the others as requested during run-time? Just an idea, I think it could work with some changes to my code I made.
That way caching would work and it would have fast-loading times still, and a smaller memory footprint unless of course u have many locales playing during run-time.
ref:
https://github.com/TrinityCore/TrinityCore/issues/22126  

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #  (insert issue tracker number)
https://github.com/TrinityCore/TrinityCore/issues/22126

**Tests performed:** (Does it build, tested in-game, etc.)
Tested in-game with a enGB client.

**Known issues and TODO list:** (add/remove lines as needed)
- Needs to be tested on non-US clients and make sure Locales are still working fine for Items, Quests, Points of interest, Creature Names, and GameObjects.

### CODE REVIEW

**Comment by Langerz82** [comment]
It didnt add the new files, and I removed them. Gonna have to do some of them again.

**Comment by Langerz82** [comment]
I got two "styles" of code the Quest templates uses an old style and arrays, the other use vectors to store the locale data and GetLocaleString. Let me know which one you think is better and I can change.

If it does not work on other client Locales its bugged and will need further fixing (I cant test on non-GB as I cant find the client).

**Comment by Langerz82** [comment]
errors bahhhh lemme fix

File: src/server/game/Server/Packets/QuestPackets.cpp [hunk 1]
```diff
@@ -23,6 +23,11 @@ void WorldPackets::Quest::QueryQuestInfo::Read()
 }
 
 WorldPacket const* WorldPackets::Quest::QueryQuestInfoResponse::Write()
+{
+    return Write(LOCALE_enUS);
```
Comment (ccrs) [inline src/server/game/Server/Packets/QuestPackets.cpp]: ?

File: src/server/game/Entities/Creature/CreatureTemplate.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+* Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
```
Comment (illfated) [inline src/server/game/Entities/Creature/CreatureTemplate.cpp]: Lines 2 through 17 should have 1 space indentation for all the Copyright headers.
I presume the code is unchanged (copied from other files), for this MaNGOS line.

**Comment by illfated** [comment]
Title suggestion: **[3.3.5] Core/Entities: Template Optimizer for Locales**

**Comment by Langerz82** [comment]
Cleaned up some of my changes a bit. Does anyone know why cs_npc.cpp is still generating that error with appveyor?

**Comment by illfated** [comment]
Sorry, I don't know. One of the real C++ coders in the TC team might know.
I can only guess it is an `#include` problem, but I could easily be wrong.

File: src/server/game/Entities/Creature/CreatureTemplate.h [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+#define TRINITYCORE_CREATURETEMPLATE_H
+
+#include "DBCEnums.h"
+#include "Common.h"
+#include "SharedDefines.h"
+
+class WorldPacket;
+namespace WorldPackets {
+    namespace Query {
+        class QueryCreatureResponse;
```
Comment (illfated) [inline src/server/game/Entities/Creature/CreatureTemplate.h]: Please follow TC C++ codestyle, curly brackets get their own separate lines in core files:
```cpp
namespace WorldPackets
{
    namespace Query
    {
        class QueryCreatureResponse;
    }
}
```

File: src/server/game/Entities/GameObject/GameObjectData.h [hunk 4]
```diff
@@ -22,558 +22,18 @@
...
-            case GAMEOBJECT_TYPE_BARBER_CHAIR: return false;
-            case GAMEOBJECT_TYPE_QUESTGIVER: return questgiver.allowMounted != 0;
-            case GAMEOBJECT_TYPE_TEXT: return text.allowMounted != 0;
-            case GAMEOBJECT_TYPE_GOOBER: return goober.allowMounted != 0;
-            case GAMEOBJECT_TYPE_SPELLCASTER: return spellcaster.allowMounted != 0;
-            default: return false;
-        }
+namespace WorldPackets {
+    namespace Query {
+        class QueryGameObjectResponse;
```
Comment (illfated) [inline src/server/game/Entities/GameObject/GameObjectData.h]: codestyle

File: src/server/game/Entities/GameObject/GameObjectTemplate.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+*
```
Comment (illfated) [inline src/server/game/Entities/GameObject/GameObjectTemplate.cpp]: codestyle

File: src/server/game/Entities/GameObject/GameObjectTemplate.h [hunk 6]
```diff
@@ -0,0 +1,586 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+*
```
Comment (illfated) [inline src/server/game/Entities/GameObject/GameObjectTemplate.h]: codestyle

File: src/server/game/Entities/Item/ItemTemplate.h [hunk 7]
```diff
@@ -21,10 +21,16 @@
...
 #include "SharedDefines.h"
-#include "WorldPacket.h"
 #include <vector>
 
 class ObjectMgr;
+class WorldPacket;
+
+namespace WorldPackets {
+    namespace Query {
+        class QueryItemSingleResponse;
```
Comment (illfated) [inline src/server/game/Entities/Item/ItemTemplate.h]: codestyle

Comment (Langerz82) [reply src/server/game/Entities/Item/ItemTemplate.h]: I'm not worried about semantics atm, the core structure of the code is the main concern but thanks for the heads up. ^_^

File: src/server/game/Entities/Creature/CreatureTemplate.h [hunk 8]
```diff
@@ -0,0 +1,175 @@
+/*
+* Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+* Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
```
Comment (illfated) [inline src/server/game/Entities/Creature/CreatureTemplate.h]: codestyle

File: src/server/game/Globals/ObjectMgr.cpp [hunk 9]
```diff
@@ -4260,21 +4258,22 @@ void ObjectMgr::LoadQuests()
...
-        typedef void(Quest::*QuestLoaderFunction)(Field* fields);
+        typedef void(Quest::*QuestLoaderFunction)(LocaleConstant lc, Field* fields);
 
         char const* QueryFields;
         char const* TableName;
         char const* TableDesc;
         QuestLoaderFunction LoaderFunction;
     };
 
+
```
Comment (illfated) [inline src/server/game/Globals/ObjectMgr.cpp]: redundant blank line with whitespace

File: src/server/game/Quests/QuestDef.h [hunk 10]
```diff
@@ -278,8 +269,8 @@ class TC_GAME_API Quest
 
         // multiple values
 
-        std::string ObjectiveText[QUEST_OBJECTIVES_COUNT];
-
+        std::string ObjectiveText[TOTAL_LOCALES][QUEST_OBJECTIVES_COUNT];
+
```
Comment (illfated) [inline src/server/game/Quests/QuestDef.h]: whitespace

**Comment by illfated** [comment]
Sorry for not being able to comment on the coding, but cleaning up the visual oddities makes it more readable for real coders.

**Comment by Seyden** [comment]
Uhm.. why did you move so much code between the files? what purpose had it?

**Comment by illfated** [comment]
Seems to be a consolidating operation, to isolate closely related code into their own files where the code seems to fit more naturally.

**Comment by Langerz82** [comment]
@Seyden it made sense to make some of them in there own Template files along with its data structures that were tightly coupled to the class.
Has anyone tested it on a non-US locale and is it working ok is the main issue.... Make sure you have emptied the cache folder in your Wow install directory so it tries to retrieve new non-US locales.

**Comment by jackpoz** [comment]
The idea of this PR sounds quite complicated to me, I would leave it simple with the choices "cache everything at start" or "don't cache anything because we don't have enough ram".

**Comment by WildCoffee** [comment]
It looks like this PR does two things. One, it splits a couple of internal-to-the-file classes into separate files. They're really only used by the classes of the files they already reside in. I don't see where this adds anything because they're not going to be used elsewhere. Two, it makes it so the server can cache multiple locale strings. I think a better approach to this would be to add a server variable that sets the default cached locale instead of defaulting to enUS. Most people connecting to a server are going to be a certain language (that's what Blizzard defines). This just seems unnecessarily complex.

**Comment by Langerz82** [comment]
1. I have to fix up the ItemTemplates to not use arrays as it's a bit of a hack.

2. Maybe we just have it so it caches the main language only or not at all. That way it's much faster doing one iteration per data item instead of 9  (I think we support atm). I agree that most Servers will be targeting one locale only.

3. Generating them on the fly is pretty damn fast i believe, since we fetch all language data from the DB and store in RAM anyway.

4. This patch was mainly trying to avoid duplicate non-locale data from being stored eating up ram. Agreed it's native data types but considering the amount of records the DB has eats quite a chunk.

**Comment by lost-illusion** [comment]
I think CACHE_DATA_QUERIES was made for "GET ME ALREADY BUILDED PACKET RIGHT NOW", 
If we cache queries by request it same as CACHE_DATA_QUERIES = 0, yes, later, we have ready cache for most popular items and performace should be ok, but... why ? We can cache, or didn't cache. 
If you have reasons for this (low RAM + long uptime), you can make for self, i mean, it dont "production" solution.
But maybe you can add this as one more config option ?
@Langerz82

**Comment by Shauren** [comment]
I still don't get the point of this PR - the config is meant to shift time spent building the packets from runtime to startup and its doing exactly that, there is nothing to fix. If you value fast startups for development then disable it

**Comment by Langerz82** [comment]
Now that I've nearly finished it I'm seeing it as being a little redundant too as you can either cache all languages or not at all. It's a nice feature just to cache only the main language you use but not really needed for the production server. Perhaps we keep the pull open and if people end up liking this feature then we re-consider it?

**Comment by illfated** [comment]
1. Why are you changing line 3 in the file `contrib/conf_merge/README` from `authserver.conf.dist` to `bnetserver.conf.dist` for the `3.3.5` branch?
2. Why are you adding `CREATE DATABASE hotfixes` to the 3.3.5 branch in `sql/create/create_mysql.sql` ?
3. Why have you removed NewLine/CR/LF at the end of the following 3 files:
    `sql/old/3.3.5a/TDB44_to_TDB45_updates/world/2011_12_13_00_world_misc.sql`
    `sql/old/3.3.5a/TDB44_to_TDB45_updates/world/2011_12_16_00_world_creature_template.sql`
    `sql/old/3.3.5a/TDB44_to_TDB45_updates/world/2011_12_16_00_world_spell_bonus_data.sql`

Much of this looks unrelated to the PR, does it really belong here?

**Comment by Langerz82** [comment]
Yep it's a mistake while using git, I'll fix soon thanks for pointing it out.

**Comment by lost-illusion** [comment]
You can add config option to config - which locales should be cached on start, one, two, or all, or nothing.
And for all non-cached locales we can use another config option - build every time, or cache for future.
And, if we cache all locales at start - why we should for every opcode search in map and etc. 
This is strange.

**Pull request closed without merging**